### PR TITLE
Extending mpi4jax with XPU support

### PR DESCRIPTION
**Pull request by jczaja**
Hi, 

This PR is a draft of changes enabling Intel XPU (via SYCL) . Changes are modeled after GPU (Cuda) integration so hopefully it should be easy to understand. 

#### Unit tests pass rate:
Currently on our setup with XPU We got:
```bash
mpirun -np 2 pytest .
============ 1 failed, 122 passed, 3 skipped, 2 warnings in 59.98s =============
```
which match result when We run on purely CPU (test_common.py I got failing for some reason on both XPU and CPU execution)

#### Usage:
New env var was added **MPI4JAX_USE_SYCL_MPI** by analogy to **MPI4JAX_USE_CUDA_MPI**.  **MPI4JAX_USE_SYCL_MPI=1** makes mpi4jax to assume that MPI implementation used can work with XPU buffers (GPU/XPU aware)  and **MPI4JAX_USE_SYCL_MPI=0**  makes mpi4jax to assume that MPI implementation cannot work XPU buffers and there is a copy of XPU buffer into allocated CPU buffer (analogically like it is done for CUDA).

#### TODO:
- Extend github action to have test for building XPU support.
- Test against different implementation of MPI (OpenMPI, IntelMPI)
- Write a doc
- Polish code

#### Concerns:
1. Collective ops (python code) of CUDA are almost the same code as SYCL (just different import of cython module) , so It would be good to make a common code for that for the ease of maintaining.
2. Cython part of XPU is using C++ (see setup.py) so perhaps It wold be good to discuss what C++ features (which standard) can be used within code e.g. Currently very minimal C++ features are used .

Any suggestions are welcomed!

@wozna, @sfraczek, @bartekkuncer, @shssf

### CODE REVIEW

**Comment by dionhaefner** [comment]
Thanks for the PR everyone. I'd like to discuss a few issues first before digging into the review.

> Collective ops (python code) of CUDA are almost the same code as SYCL (just different import of cython module) , so It would be good to make a common code for that for the ease of maintaining.

Are the descriptors different from the CUDA ones?

> Cython part of XPU is using C++ (see setup.py) so perhaps It wold be good to discuss what C++ features (which standard) can be used within code e.g. Currently very minimal C++ features are used .

What's the rationale for this? Is that because SYCL doesn't have a C interface?

Also a nit: Please apply the pre-commit hooks to all code in the PR to ensure a compatible formatting with the rest of the codebase.

**Comment by sfraczek** [comment]
Hi,
The descriptors are identical just xpu->gpu and cuda->sycl.
Sycl doesn't have C interface. I personally don't think for now we have to worry about defining C++ features upfront.
Mostly the difference in code in bridge/ops can be summarized as:
- xstream is casted to `sycl::queue` and we use two methods `queue.wait()` and `queue.memcpy()` which can throw.
- `checked_cuda_stream_synchronize` -> `checked_sycl_queue_wait`
- `checked_cuda_memcpy` -> `checked_sycl_memcpy`

File: mpi4jax/_src/collective_ops/allreduce.py
Comment (dionhaefner) [inline mpi4jax/_src/collective_ops/allreduce.py:281]:
```suggestion
mlir.register_lowering(mpi_allreduce_p, translation_rule_gpu(mpi_allreduce_xla_encode_gpu), platform="cuda")
mlir.register_lowering(mpi_allreduce_p, translation_rule_xpu(mpi_allreduce_xla_encode_gpu), platform="xpu")
```

This should work, right? Since the descriptors are the same structs anyway, and the kernels have the same name?

Comment (sfraczek) [reply mpi4jax/_src/collective_ops/allreduce.py:281]: I am working on it and I will propose something because there are slight changes

**Comment by PhilipVinc** [comment]
Hi @jczaja , what is the status with the PR? is it somewhat in a reviewable state? What is missing?

**Comment by jczaja** [comment]
> Hi @jczaja , what is the status with the PR? is it somewhat in a reviewable state? What is missing?
1. It is still work in progress. 
2.  @dionhaefner gave some suggestions that we are working on. No need to review now as the branch with changes keep changing freqently.
3. Missing (not finished) items:
   - Documentation (WIP)
   - Refactoring and polishing according to @dionhaefner suggestions (WIP)
   - Testing against various Intel XPU/GPU and MPI implementations (WIP)

One we have it done I will change Draft to Pull request and ask you to review.

File: mpi4jax/_src/collective_ops/allreduce.py
Comment (sfraczek) [reply mpi4jax/_src/collective_ops/allreduce.py:254]: The files `mpi_xla_bridge_gpu.pyx` and `mpi_xla_bridge_xpu.pyx` aren't exactly identical. The differences are quite small. I don't know how to reduce duplication in there.
I tried to get close to your suggestion by writing it like this:
```python
mlir.register_lowering(
    mpi_allreduce_p,
    translation_rule_xpu(
        partial(
            mpi_allreduce_xla_encode_device,
            build_allreduce_descriptor=xla_bridge.mpi_xla_bridge_xpu.build_allreduce_descriptor,
        )
    ),
    platform="xpu",
)
``` 
but it doesn't allow me to use module mpi_xla_bridge_gpu because it doesn't exist when installing for xpu only.

Comment (dionhaefner) [reply mpi4jax/_src/collective_ops/allreduce.py:254]: >  build_allreduce_descriptor=xla_bridge.mpi_xla_bridge_xpu.build_allreduce_descriptor

I don't get that one, isn't the XPU descriptor exactly the same as the GPU descriptor?

**Comment by jczaja** [comment]
@dionhaefner , @PhilipVinc 
This PR is ready to be reviewed now. One concern is that codecov CI is failing and I'm not sure if we can do something about as unit tests are not running on XPU neither on GPU in this github CI. So we added more code that is not covered in github CI.

Anyway , We are looking forward to your feedback.

File: setup.py
Comment (dionhaefner) [inline setup.py:130]: Is this the only way to get the path?

File: setup.py
Comment (dionhaefner) [inline setup.py:11]: Let's remove this + other calls to logging

File: setup.py
Comment (dionhaefner) [inline setup.py:227]: What does this do?

File: setup.py
Comment (dionhaefner) [inline setup.py:232]: Let's be more neutral here, this typically doesn't imply a problem.

```suggestion
            "SYCL (Intel Basekit) path not found",
```

**Review by dionhaefner — CHANGES_REQUESTED** [review]
Thanks! This looks high-quality, let's fix some minor issues.

In case this breaks in the future: Who is going to maintain this code?

File: setup.py
Comment (jczaja) [reply setup.py:227]: > What does this do?

- `languages="c++"` Sycl is a C++ API so we have to use C++ compiler to build xpu cython files

- `define_macros=[("OMPI_SKIP_MPICXX", "1")]` this instructs C++ compiler to ignore OpenMPI C++ bindings. This came up during testing of this PR across diffrent MPI implementations. MPI used to have C++ bindings that are now deprecated. However on some linux distros (Fedora 36) OpenMPI comes with enabled C++ bindings. For example on Fedora 36: `ompi_info | grep "C++ binding"` produces: `C++ bindings: yes` . At the beginning we have bindings enabled but there were problems with missing symbols from OpenMPI C++ bindings module and similar problems were encountered by various people on internet . So we decided to switch them off as C++ bindings are deprecated anyway. And this macro `OMPI_SKIP_MPICXX` set to `1` is telling C++ compiler to ignore those bindings.

File: setup.py
Comment (jczaja) [reply setup.py:130]: You need SYCL lib and headers to build XPU extension of mpi4jax. As shown in yml file for testing building of XPU extension we can call  `source $HOME/basekit/setvars.sh`
which is the way to enable Intel oneAPI basekit. By the default it will enable all installed components like Intel components, sycl, MPI etc.  And this process will also set ONEAPI_ROOT var. In theory you could export manually ONEAPI_ROOT env var without calling `setvars.sh` as an alternative solution.

Comment (PhilipVinc) [reply setup.py:130]: For cuda, just a few lines above, if `CUDA_ROOT` is not set we test if `/usr/local/cuda` exists and/or if `nvcc` can be used to get the root.

Maybe it could be possible to try some standard default path and check if it installed there?

**Comment by jczaja** [comment]
> Thanks! This looks high-quality, let's fix some minor issues.
> 
> In case this breaks in the future: Who is going to maintain this code?

@dionhaefner Thanks for reviewing this PR.  This functionality will be maintained by Intel JAX GPU teams which is: @jczaja , @sfraczek , @Silv3S , @tsocha, @shssf and @wozna. If in the future we are reassigned to other projects then we will let you know about new maintainers.

We started to work on suggestions you made and will update PR as soon as we have something.

**Comment by jczaja** [comment]
> Thanks! This looks high-quality, let's fix some minor issues.
> 
> In case this breaks in the future: Who is going to maintain this code?

@dionhaefner Thanks for reviewing this PR.  This functionality will be maintained by Intel JAX GPU teams . I will send some more info via email

We started to work on suggestions you made and will update PR as soon as we have something.

File: setup.py
Comment (jczaja) [reply setup.py:130]: We discussed this internally and ```ONEAPI_ROOT``` is having an default location : ```/opt/intel/oneapi``` , but inside of this there maybe a couple of compilers(this is where SYCL is). Used compiler  is pointed by  ```CMPLR_ROOT``` env var so we will update PR with using this env var instead of ```ONEAPI_ROOT```.

In case of fallback w.g. no ```CMPLR_ROOT``` is set then we could fallback to most recent SYCL compiler from default location.

File: setup.py
Comment (dionhaefner) [reply setup.py:227]: Okay thanks. Could you add that explanation as a comment in the code please?

**Comment by jczaja** [comment]
@dionhaefner , @PhilipVinc We adressed (hopefully) all the questions and comments and updated PR with relevant changes. So we are ready for second round of review. Please review!

File: mpi4jax/_src/collective_ops/allgather.py
Comment (dionhaefner) [inline mpi4jax/_src/collective_ops/allgather.py:185]:
```suggestion
mpi_allgather_xla_encode_xpu = translation_rule_xpu(mpi_allgather_xla_encode_device)
```

File: mpi4jax/_src/xla_bridge/device_descriptors.pxd
Comment (dionhaefner) [inline mpi4jax/_src/xla_bridge/device_descriptors.pxd:9]:
```suggestion
from mpi4py.libmpi cimport (
    MPI_Comm,
    MPI_Datatype,
    MPI_Op,
    MPI_Status,
)
```

File: mpi4jax/_src/xla_bridge/device_descriptors.pyx
Comment (dionhaefner) [inline mpi4jax/_src/xla_bridge/device_descriptors.pyx:11]:
```suggestion
from mpi4py.libmpi cimport (
    MPI_Comm,
    MPI_Datatype,
    MPI_Op,
    MPI_Status,
)
```

File: mpi4jax/_src/xla_bridge/mpi_xla_bridge_gpu.pyx
Comment (dionhaefner) [inline mpi4jax/_src/xla_bridge/mpi_xla_bridge_gpu.pyx:29]:
```suggestion
from .device_descriptors cimport (
    RecvDescriptor,
    AllgatherDescriptor,
    AllreduceDescriptor,
    AlltoallDescriptor,
    BarrierDescriptor,
    BcastDescriptor,
    GatherDescriptor,
    ReduceDescriptor,
    ScanDescriptor,
    ScatterDescriptor,
    SendDescriptor,
    SendrecvDescriptor,
)
```

File: mpi4jax/_src/xla_bridge/mpi_xla_bridge_xpu.pyx
Comment (dionhaefner) [inline mpi4jax/_src/xla_bridge/mpi_xla_bridge_xpu.pyx:31]:
```suggestion
from .device_descriptors cimport (
    RecvDescriptor,
    AllgatherDescriptor,
    AllreduceDescriptor,
    AlltoallDescriptor,
    BarrierDescriptor,
    BcastDescriptor,
    GatherDescriptor,
    ReduceDescriptor,
    ScanDescriptor,
    ScatterDescriptor,
    SendDescriptor,
    SendrecvDescriptor,
)
```

File: mpi4jax/_src/xla_bridge/mpi_xla_bridge_xpu.pyx
Comment (dionhaefner) [inline mpi4jax/_src/xla_bridge/mpi_xla_bridge_xpu.pyx:68]:
```suggestion
        mpi_xla_bridge.abort(0, comm, "Error: Unable to copy Data from the Device to CPU")
```

File: mpi4jax/_src/xla_bridge/mpi_xla_bridge_xpu.pyx
Comment (dionhaefner) [inline mpi4jax/_src/xla_bridge/mpi_xla_bridge_xpu.pyx:75]:
```suggestion
        mpi_xla_bridge.abort(0, comm, "Error: Unable to execute SYCL queue::wait()")
```

File: mpi4jax/_src/xla_bridge/mpi_xla_bridge_xpu.pyx
Comment (dionhaefner) [inline mpi4jax/_src/xla_bridge/mpi_xla_bridge_xpu.pyx:64]: Why is this holding the GIL?

File: mpi4jax/_src/xla_bridge/sycl_runtime_api.pxd
Comment (dionhaefner) [inline mpi4jax/_src/xla_bridge/sycl_runtime_api.pxd:7]:
```suggestion
```

File: mpi4jax/_src/xla_bridge/sycl_runtime_api.pxd
Comment (dionhaefner) [inline mpi4jax/_src/xla_bridge/sycl_runtime_api.pxd:11]:
```suggestion
```

File: mpi4jax/experimental/notoken/collective_ops/allgather.py
Comment (dionhaefner) [inline mpi4jax/experimental/notoken/collective_ops/allgather.py:29]:
```suggestion
from mpi4jax._src.xla_bridge.device_descriptors import build_allgather_descriptor
```

File: setup.py
Comment (dionhaefner) [inline setup.py:132]: Please mirror the corresponding CUDA version by checking whether paths exist (return `None` otherwise)

File: setup.py
Comment (dionhaefner) [inline setup.py:143]: Is OSX supported?

File: setup.py
Comment (dionhaefner) [inline setup.py:225]:
```suggestion
                # This macro instructs C++ compiler to ignore potential existence of
```

File: docs/installation.rst
Comment (dionhaefner) [inline docs/installation.rst:91]: Note to self: give these docs another pass before merging

**Review by dionhaefner — CHANGES_REQUESTED** [review]

**Comment by dionhaefner** [comment]
Thanks, I like the reduced code duplication. Let's get some final changes done then merge this.

File: mpi4jax/_src/collective_ops/allgather.py
Comment (dionhaefner) [reply mpi4jax/_src/collective_ops/allgather.py:185]: (same for GPU and other translation rules)

File: mpi4jax/experimental/notoken/collective_ops/allgather.py
Comment (dionhaefner) [reply mpi4jax/experimental/notoken/collective_ops/allgather.py:29]: (same for other notoken modules)

**Comment by dionhaefner** [comment]
@jczaja Could you please address the last 3 comments so we can get this merged?

**Comment by Zantares** [comment]
> @jczaja Could you please address the last 3 comments so we can get this merged?

Hi @dionhaefner , we have taken related work and will give the feedback in 1 week, thanks for your patience!

File: setup.py
Comment (Dboyqiao) [reply setup.py:146]: No. Intel oneAPI Toolkits will not support macOS since the 2024.0 release.

File: mpi4jax/_src/xla_bridge/mpi_xla_bridge_xpu.pyx
Comment (Dboyqiao) [reply mpi4jax/_src/xla_bridge/mpi_xla_bridge_xpu.pyx:63]: we use try-except statement to catch the exception from sycl queue memcpy in this function, which requires the GIL. Otherwise, it will be failed as "Try-except statement not allowed without gil" when compling this cython file.

File: setup.py
Comment (Dboyqiao) [reply setup.py]: updated

**Comment by Dboyqiao** [comment]
> @jczaja Could you please address the last 3 comments so we can get this merged?

@dionhaefner we have addressed the left 3 comments. could you help to review？ Thank you!

File: mpi4jax/_src/xla_bridge/mpi_xla_bridge_xpu.pyx
Comment (dionhaefner) [reply mpi4jax/_src/xla_bridge/mpi_xla_bridge_xpu.pyx:63]: OK, so it's to catch the C++ exception - makes sense.

**Review by dionhaefner — APPROVED** [review]
LGTM, thanks all!

@PhilipVinc You want to give it a final pass?

**Pull request merged**
